L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+  // simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEback")))
+  dummyClustering_(_config.getParameterSet("dummyClustering"))
+{
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>("HGCalBackendLayer1Processor2DClustering");
+  produces<l1t::HGCalMulticlusterBxCollection>("HGCalBackendLayer2Processor3DClustering");
+  if (makeCellsCollection_)
+    produces<l1t::HGCalTriggerCellBxCollection>("HGCalVFEProcessorSums");
+}
+
+CaloTruthCellsProducer::~CaloTruthCellsProducer()
+{
+}
+
+void
+CaloTruthCellsProducer::produce(edm::Event& _event, edm::EventSetup const& _setup)
+{
+  edm::Handle<CaloParticleCollection> caloParticlesHandle;
+  _event.getByToken(caloParticlesToken_, caloParticlesHandle);
+  auto& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  _event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto& triggerCells(*triggerCellsHandle);
+
+  // std::map<uint32_t, double> hitMap(makeHitMap(_event, _setup)); // cellId -> sim energy
+  // std::map<uint32_t, double> tcMap; // tcId -> total sim energy
+
+  dummyClustering_.eventSetup(_setup);
+  showerShape_.eventSetup(_setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  _setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto& geometry(*geometryHandle);
+
+  std::map<uint32_t, CaloParticleRef> tcToCalo;
+
+  // used later to order multiclusters
+  std::vector<CaloParticleRef> orderedCaloRefs;
+
+  for (unsigned iP(0); iP != caloParticles.size(); ++iP) {
+    auto& caloParticle(caloParticles.at(iP));
+    if (caloParticle.g4Tracks().at(0).eventId().event() != 0) // pileup
+      continue;
+
+    CaloParticleRef ref(caloParticlesHandle, iP);
+
+    SimClusterRefVector const& simClusters(caloParticle.simClusters());
+    for (auto& simCluster : simClusters) {
+      for (auto& hAndF : simCluster->hits_and_fractions()) {
+        DetId hitId(hAndF.first);
+        HGCalDetId tcId;
+        try {
+          tcId = HGCalDetId(geometry.getTriggerCellFromCell(hitId));
+        }
+        catch (cms::Exception const& ex) {
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        // tcMap[tcId] += hitMap[hitId.rawId()] * hAndF.first;
+        tcToCalo.emplace(tcId, ref);
+      }
+    }
+
+    // ordered by the gen particle index
+    int genIndex(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    if (genIndex < 0) // shouldn't happen
+      continue;
+    if (genIndex >= int(orderedCaloRefs.size()))
+      orderedCaloRefs.resize(genIndex + 1);
+
+    orderedCaloRefs[genIndex] = ref;
+  }
+
+  auto outMap(std::make_unique<CaloToCellsMap>(caloParticlesHandle, triggerCellsHandle));
+  std::unique_ptr<l1t::HGCalTriggerCellBxCollection> outCollection;
+  if (makeCellsCollection_)
+    outCollection.reset(new l1t::HGCalTriggerCellBxCollection);
```

Comment:
Better to use `make_unique`

Reply:
Why?

Reply:
1) https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rh-make_unique
2) because I know it will be a comment we will have when integrating this in the official CMSSW

Reply:
OK.. changed to
```
if (makeCellsCollection_)
  outCollection = std::move(std::make_unique<l1t::HGCalTriggerCellBxCollection>());
```
but not sure if that's what you wanted.